gtsfm/utils/homography_decomposition.py

```diff
@@ -0,0 +1,384 @@
+See how it is used in COLMAP:
+https://github.com/colmap/colmap/blob/dev/src/estimators/two_view_geometry.cc#L198
+
+COLMAP and OpenCV's implementationsare based off of:
+Ezio Malis, Manuel Vargas, and others. Deeper understanding of the homography decomposition for vision-based control. 2007.
+https://hal.inria.fr/inria-00174036/PDF/RR-6303.pdf
+
+OpenCV does not support the case of intrinsics from two separate cameras, however.
+
+Authors: John Lambert (Python), from original C++
+"""
+
+from typing import List, Tuple
+
+import numpy as np
+from gtsam import Rot3, Unit3
+
+"""
+PoseFromHomographyMatrix(
+        H, camera1.CalibrationMatrix(), camera2.CalibrationMatrix(),
+        inlier_points1_normalized, inlier_points2_normalized, &R, &tvec, &n,
+        &points3D);
+
+  if (points3D.empty()) {
+    tri_angle = 0;
+  } else {
+    tri_angle = Median(CalculateTriangulationAngles(
+        Eigen::Vector3d::Zero(), -R.transpose() * tvec, points3D));
+  }
+
+  if (config == PLANAR_OR_PANORAMIC) {
+    if (tvec.norm() == 0) {
+      config = PANORAMIC;
+      tri_angle = 0;
+    } else {
+      config = PLANAR;
+    }
+"""
+
+
+def pose_from_homography_matrix(
+    H: np.ndarray,
+    K1: np.ndarray,
+    K2: np.ndarray,
+    points1: np.ndarray,
+    points2: np.ndarray,
+) -> Tuple[Rot3, Unit3, np.ndarray, np.ndarray]:
+    """Recover the most probable pose from the given homography matrix.
+
+    Args:
+        H: array of shape (3,3)
+        K1: array of shape (3,3) representing camera 1's intrinsics
+        K2: array of shape (3,3) representing camera 2's intrinsics
+        points1: array of shape (N,2)
+        points2: array of shape (N,2)
+
+    Returns:
+        R: relative rotation matrix.
+        t: translation direction.
+        n: array of shape (3,) representing plane normal vector.
+        points3D: array of shape (N,3) representing triangulated 3d points.
+    """
+    if points1.shape != points2.shape:
+        raise RuntimeError("Coordinates of 2d correspondences must have the same shape.")
+
+    R_cmbs, t_cmbs, n_cmbs = decompose_homography_matrix(H, K1, K2)
+
+    for i in range(len(R_cmbs)):
+        points3D_cmb = check_cheirality(R_cmbs[i], t_cmbs[i], points1, points2)
+        if len(points3D_cmb) >= len(points3D):
+            R = R_cmbs[i]
+            t = t_cmbs[i]
+            n = n_cmbs[i]
+            points3D = points3D_cmb
+
+    return R, t, n, points3D
+
+
+def check_cheirality(R: Rot3, t: np.ndarray, points1: np.ndarray, points2: np.ndarray) -> np.ndarray:
+    """
+    Args:
+        R: array of shape (3,3)
+        t: array of shape (3,)
+        points1:
+        points2:
+
+    Returns:
+        points3D: array of shape (N,3)
+    """
+    if points1.shape != points2.shape:
+        raise RuntimeError("Coordinates of 2d correspondences must have the same shape.")
+
+    # try triangulating each point
+
+    camera_dict = {0: PinholeCameraCal3Bundler(), 1: PinholeCameraCal3Bundler()}
```

Comment:
both cameras are exactly the same?

Reply:
fixed now, they have different poses now in the code